extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+  onAmpCreativeShadowDomRender() {}
+
+  /**
+   * Send ad request, extract the creative and signature from the response.
+   * @param {string} adUrl Request URL to send XHR to.
+   * @return {!Promise<?FetchResponse>}
+   * @private
+   */
+  sendXhrRequest_(adUrl) {
+    const xhrInit = {
+      mode: 'cors',
+      method: 'GET',
+      credentials: 'include',
+      // TODO(kjwright):  Add requireAmpResponseSourceOrigin once supported
+      // server-side
+    };
+    return xhrFor(this.getWin())
+        .fetch(adUrl, xhrInit)
+        .catch(unusedReason => {
+          // Error so set rendered_ so iframe will not be written on
+          // layoutCallback.
+          // TODO: is this the appropriate action?  Perhaps we should just allow
+          // the ad to be rendered via iframe.
+          this.rendered_ = true;
+          return null;
+        });
+  }
+
+  /**
+   * Try to validate creative is AMP through crypto signature.
+   * @param {!FetchResponse} fetchResponse
+   * @param {!ArrayBuffer} bytes
+   * @return {!Promise<boolean>}
+   * @private
+   */
+  validateAdResponse_(fetchResponse, bytes) {
+    return this.extractCreativeAndSignature(bytes, fetchResponse.headers)
+        .then(adResponse => {
+          // Validate when we have a signature and we have native crypto.
+          if (adResponse.signature && verifySignatureIsAvailable()) {
+            try {
+              // Among other things, the signature might not be proper base64.
+              return verifySignature(adResponse.creativeArrayBuffer,
+                  base64ToByteArray(adResponse.signature), publicKeyInfos);
+            } catch (e) {}
+          }
+          return false;
+        });
+  }
+
+  /**
+   * Render the validated AMP creative directly in the parent page.
+   * @param {boolean} valid If the ad response signature was valid.
+   * @return {Promise<boolean>} Whether the creative was successfully
+   *     rendered.
+   * @private
+   */
+  maybeRenderAmpAd_(valid, bytes) {
+    if (!valid) {
+      return Promise.resolve(false);
+    }
+    // Timer id will be set if we have entered layoutCallback at which point
+    // 3p throttling count was incremented.  We want to "release" the throttle
+    // immediately since we now know we are not a 3p ad.
+    if (this.timerId_) {
+      decrementLoadingAds(this.timerId_, this.getWin());
+    }
+    return xhrFor(this.getWin()).utf8FromArrayBuffer(bytes).then(creative => {
+      // Find the json blob located at the end of the body and parse it.
+      const creativeMetaData = this.getAmpAdMetadata_(creative);
+      if (!creativeMetaData || !this.supportsShadowDom()) {
+        // Shadow DOM is not supported or could not find appropriate markers
+        // within the creative therefore load within cross domain iframe.
+        // Iframe is created immediately (as opposed to waiting for
+        // layoutCallback) as the the creative has been verified as AMP and
+        // will run efficiently.
```

Comment:
It will run efficiently, but how much lag it going to be introduced form creating and injecting an iframe? This is outside out normal Task Queueing system, and we won't be able to compensate for slow frame rates (our Task Queue limits itself to a frame only).